frontend/packages/console-dynamic-plugin-sdk/src/utils/fetch/console-fetch.ts

```diff
@@ -32,17 +32,30 @@ const consoleFetchCommon = async (
   method: string = 'GET',
   options: RequestInit = {},
   timeout?: number,
+  isFullResponse?: boolean,
 ) => {
   const headers = getConsoleRequestHeaders();
+  const defaultIsFullResponse = typeof isFullResponse !== 'boolean' ? false : isFullResponse;
+
   // Pass headers last to let callers to override Accept.
   const allOptions = _.defaultsDeep({ method }, options, { headers });
   const response = await consoleFetch(url, allOptions, timeout);
   const text = await response.text();
   const isPlainText = response.headers.get('Content-Type') === 'text/plain';
   if (!text) {
-    return isPlainText ? '' : {};
+    return isPlainText ? response : {};
   }
-  return isPlainText || !response.ok ? text : JSON.parse(text);
+
+  // Map cosole respose to data, headers and status
+  const fullResponse: { data: Body; headers: { [key: string]: any }; status: number } = {
+    headers: [...response.headers].reduce((acc, header) => {
+      return { ...acc, [header[0]]: header[1] };
+    }, {}),
+    status: response.status,
+    data: JSON.parse(text),
+  };
+
+  return !response.ok ? text : defaultIsFullResponse ? fullResponse : fullResponse.data;
```

Comment:
> If we factored out a helper it would simplify this a lot:
> 
> ```ts
> const parseData = (response) => {
>   const text = await response.text();
>   const isPlainText = response.headers.get('Content-Type') === 'text/plain';
>   if (!text) {
>     return isPlainText ? '' : {};
>   }
>   return isPlainText || !response.ok ? text : JSON.parse(text);
> }
> 
> const consoleFetchCommon = async (
>   url: string,
>   method: string = 'GET',
>   options: RequestInit = {},
>   timeout?: number,
>   preserveResponse?: boolean,
> ) => {
>   const headers = getConsoleRequestHeaders();
>   const allOptions = _.defaultsDeep({ method }, options, { headers });
>   const response = await consoleFetch(url, allOptions, timeout);
>   return preserveResponse ? response : parseData(response);
> };
> ```

@TheRealJon The `parseData` logic  looks small and it is used in one place only, so that is why I did not extract it into another function

Reply:
IMO, size and reusability are not the only criteria to consider when factoring out a helper. In this case, it helps readability.

Reply:
IMO, size and reusability are not the only criteria to consider when factoring out a helper. In this case, it helps readability.